File: public/app/pages/configuration.js
Vue.component('Configuration', {
	template: '{{include=template/pages/configuration.html}}',
	data: function(){
		return {
			configs: [],
			tracks: [],
			cars: [],
			activePaintings: [],
			selectedTrack: 0,
			selectedCar: 0,
			selectedPainting: 0,
			spectator: false,
			driver: '',
			team: '',
			guid: '',
			fixed_setup: '',
			// ---
			selectedCars: [],
			weather: [],
			// ---
			_id: 0,
			name: 'Servername',
			pwd: '',
			admin_pwd: '',
			pickup_mode: true,
			lock_entry_list: false,
			race_overtime: 60,
			max_slots: 0,
			result_screen_time: 60,
			welcome: '',
			description: '',
			udp: 9600,
			tcp: 9600,
			http: 8081,
			packets_hz: 18,
			loop_mode: true,
			show_in_lobby: true,
			threads: 2,
			abs: 1,
			tc: 1,
			stability_aid: false,
			auto_clutch: false,
			tyre_blankets: true,
			force_virtual_mirror: true,
			fuel_rate: 100,
			damage_rate: 50,
			tires_wear_rate: 100,
			allowed_tires_out: 2,
			max_ballast: 150,
			disable_gas_cut_penality: false,
			dynamic_track: true,
			condition: 'CUSTOM',
			start_value: 100,
			randomness: 0,
			transferred_grip: 100,
			laps_to_improve_grip: 1,
			kick_vote_quorum: 70,
			session_vote_quorum: 70,
			vote_duration: 15,
			blacklist: 0,
			max_collisions_km: 5,
			booking: false,
			booking_time: 0,
			practice: true,
			practice_time: 15,
			can_join_practice: true,
			qualify: true,
			qualify_time: 15,
			can_join_qualify: true,
			race: true,
			race_laps: 10,
			race_time: 0,
			race_wait_time: 60,
			race_extra_lap: false,
			join_type: 1,
			time: '14:00',
			sun_angle: 16,
			track: '',
			legal_tyres: '',
			udp_plugin_local_port: 0,
			udp_plugin_address: '',
			race_pit_window_start: 0,
			race_pit_window_end: 0,
			reversed_grid_race_positions: 0,
			// ---
			err: 0,
			addEditConfig: false,
			removeConfig: false,
			saved: false,
			removed: false
		}
	},
	mounted: function(){
		this._load();
	},
	watch: {
		condition: function (value) {
			this.populateDynamicTrackWithPreset(value);
		}
	},
	methods: {
		_load: function(){
			this.$http.get('/api/configuration')
			.then(function(resp){
				if(resp.data.code){
					console.log(resp.data.code+': '+resp.data.msg);
					return;
				}

				this.configs = resp.data;
			});

			this.$http.get('/api/tracks')
			.then(function(resp){
				if(resp.data.code){
					console.log(resp.data.code+': '+resp.data.msg);
					return;
				}

				this.tracks = resp.data;
			});

			this.$http.get('/api/cars')
			.then(function(resp){
				if(resp.data.code){
					console.log(resp.data.code+': '+resp.data.msg);
					return;
				}

				this.cars = resp.data;

				if(this.cars.length){
					this.activePaintings = this.cars[0].paintings;
				}
			});
		},
		_reset: function(){
			this.spectator = false;
			this.driver = '';
			this.team = '';
			this.guid = '';
			this.fixed_setup = '';

			this.selectedCars = [];
			this.weather = [];

			this._id = 0;
			this.name = 'Servername';
			this.pwd = '';
			this.admin_pwd = '';
			this.pickup_mode = true;
			this.lock_entry_list = false;
			this.race_overtime = 60;
			this.max_slots = 0;
			this.result_screen_time = 60;
			this.welcome = '';
			this.description = '';
			this.udp = 9600;
			this.tcp = 9600;
			this.http = 8081;
			this.packets_hz = 18;
			this.loop_mode = true;
			this.show_in_lobby = true;
			this.threads = 2;
			this.abs = 1;
			this.tc = 1;
			this.stability_aid = false;
			this.auto_clutch = false;
			this.tyre_blankets = true;
			this.force_virtual_mirror = true;
			this.fuel_rate = 100;
			this.damage_rate = 50;
			this.tires_wear_rate = 100;
			this.allowed_tires_out = 2;
			this.max_ballast = 150;
			this.disable_gas_cut_penality = false;
			this.dynamic_track = true;
			this.condition = 'CUSTOM';
			this.start_value = 100;
			this.randomness = 0;
			this.transferred_grip = 100;
			this.laps_to_improve_grip = 1;
			this.kick_vote_quorum = 70;
			this.session_vote_quorum = 70;
			this.vote_duration = 15;
			this.blacklist = 0;
			this.max_collisions_km = 5;
			this.booking = false;
			this.booking_time = 0;
			this.practice = true;
			this.practice_time = 15;
			this.can_join_practice = true;
			this.qualify = true;
			this.qualify_time = 15;
			this.can_join_qualify = true;
			this.race = true;
			this.race_laps = 10;
			this.race_time = 0;
			this.race_wait_time = 60;
			this.race_extra_lap = false;
			this.join_type = 1;
			this.time = '14:00';
			this.sun_angle = 16;
			this.track = '';
			this.legal_tyres = '';
			this.udp_plugin_local_port = 0;
			this.udp_plugin_address = '';
			this.race_pit_window_start = 0;
			this.race_pit_window_end = 0;
			this.reversed_grid_race_positions = 0;
			
			this.err = 0;
			this.addEditConfig = false;
			this.removeConfig = false;
			this.saved = false;
			this.removed = false;
		},
		_openConfig: function(id, copy){
			this.$http.get('/api/configuration', {params: {id: id}})
			.then(function(resp){
				if(resp.data.code){
					console.log(resp.data.code+': '+resp.data.msg);
					return;
				}

				// config
				this.name = resp.data.name;
				this.pwd = resp.data.pwd;
				this.admin_pwd = resp.data.admin_pwd;
				this.pickup_mode = resp.data.pickup_mode;
				this.lock_entry_list = resp.data.lock_entry_list;
				this.race_overtime = resp.data.race_overtime;
				this.max_slots = resp.data.max_slots;
				this.welcome = resp.data.welcome;
				this.description = resp.data.description;
				this.udp = resp.data.udp;
				this.tcp = resp.data.tcp;
				this.http = resp.data.http;
				this.packets_hz = resp.data.packets_hz;
				this.loop_mode = resp.data.loop_mode;
				this.show_in_lobby = resp.data.show_in_lobby;
				this.threads = resp.data.threads;
				this.abs = resp.data.abs;
				this.tc = resp.data.tc;
				this.stability_aid = resp.data.stability_aid;
				this.auto_clutch = resp.data.auto_clutch;
				this.tyre_blankets = resp.data.tyre_blankets;
				this.force_virtual_mirror = resp.data.force_virtual_mirror;
				this.fuel_rate = resp.data.fuel_rate;
				this.damage_rate = resp.data.damage_rate;
				this.tires_wear_rate = resp.data.tires_wear_rate;
				this.allowed_tires_out = resp.data.allowed_tires_out;
				this.max_ballast = resp.data.max_ballast;
				this.disable_gas_cut_penality = resp.data.disable_gas_cut_penality;
				this.result_screen_time = resp.data.result_screen_time;
				this.dynamic_track = resp.data.dynamic_track;
				this.condition = resp.data.condition;
				this.start_value = resp.data.start_value;
				this.randomness = resp.data.randomness;
				this.transferred_grip = resp.data.transferred_grip;
				this.laps_to_improve_grip = resp.data.laps_to_improve_grip;
				this.kick_vote_quorum = resp.data.kick_vote_quorum;
				this.session_vote_quorum = resp.data.session_vote_quorum;
				this.vote_duration = resp.data.vote_duration;
				this.blacklist = resp.data.blacklist;
				this.max_collisions_km = resp.data.max_collisions_km;
				this.booking = resp.data.booking;
				this.booking_time = resp.data.booking_time;
				this.practice = resp.data.practice;
				this.practice_time = resp.data.practice_time;
				this.can_join_practice = resp.data.can_join_practice;
				this.qualify = resp.data.qualify;
				this.qualify_time = resp.data.qualify_time;
				this.can_join_qualify = resp.data.can_join_qualify;
				this.race = resp.data.race;
				this.race_laps = resp.data.race_laps;
				this.race_time = resp.data.race_time;
				this.race_wait_time = resp.data.race_wait_time;
				this.race_extra_lap = resp.data.race_extra_lap;
				this.join_type = resp.data.join_type;
				this.time = resp.data.time;
				this.sun_angle = resp.data.sun_angle;
				this.legal_tyres = resp.data.legal_tyres;
				this.udp_plugin_local_port = resp.data.udp_plugin_local_port;
				this.udp_plugin_address = resp.data.udp_plugin_address;
				this.race_pit_window_start = resp.data.race_pit_window_start;
				this.race_pit_window_end = resp.data.race_pit_window_end;
				this.reversed_grid_race_positions = resp.data.reversed_grid_race_positions;

				if(copy){
					this.name += ' (copy)';
				}

				// track
				for(var i = 0; i < this.tracks.length; i++){
					if(this.tracks[i].name == resp.data.track && this.tracks[i].config == resp.data.track_config){
						this.selectTrack(i);
						break;
					}
				}
				
				// weather
				this.weather = resp.data.weather;

				if(copy){
					for(var i = 0; i < this.weather.length; i++){
						this.weather[i].id = 0;
					}
				}

				// cars
				this.selectedCars = resp.data.cars;

				if(copy){
					for(var i = 0; i < this.selectedCars.length; i++){
						this.selectedCars[i].id = 0;
					}
				}

				this.addEditConfig = true;
			});
		},
		openAddEditConfig: function(id){
			this._reset();
			
			if(id){
				this._id = id;
				this._openConfig(id, false);
			}
			else{
				this.addEditConfig = true;
			}
		},
		openRemoveConfig: function(id){
			this._reset();

			if(!id){
				return;
			}

			this._id = id;
			this.removeConfig = true;
		},
		copyConfig: function(id){
			this._openConfig(id, true);
		},
		performAddEditConfig: function(){
			for(var i = 0; i < this.weather.length; i++){
				this.weather[i].base_ambient_temp = parseInt(this.weather[i].base_ambient_temp);
				this.weather[i].realistic_road_temp = parseInt(this.weather[i].realistic_road_temp);
				this.weather[i].base_road_temp = parseInt(this.weather[i].base_road_temp);
				this.weather[i].ambient_variation = parseInt(this.weather[i].ambient_variation);
				this.weather[i].road_variation = parseInt(this.weather[i].road_variation);
				this.weather[i].wind_base_speed_min = parseInt(this.weather[i].wind_base_speed_min);
				this.weather[i].wind_base_speed_max = parseInt(this.weather[i].wind_base_speed_max);
				this.weather[i].wind_base_direction = parseInt(this.weather[i].wind_base_direction);
				this.weather[i].wind_variation_direction = parseInt(this.weather[i].wind_variation_direction);
			}

			for(var i = 0; i < this.selectedCars.length; i++){
				this.selectedCars[i].position = i;
			}

			var data = {
				id: this._id,
				name: this.name,
				pwd: this.pwd,
				admin_pwd: this.admin_pwd,
				pickup_mode: this.pickup_mode,
				lock_entry_list: this.lock_entry_list,
				race_overtime: parseInt(this.race_overtime),
				max_slots: parseInt(this.max_slots),
				welcome: this.welcome,
				description: this.description,
				udp: parseInt(this.udp),
				tcp: parseInt(this.tcp),
				http: parseInt(this.http),
				packets_hz: parseInt(this.packets_hz),
				loop_mode: this.loop_mode,
				show_in_lobby: this.show_in_lobby,
				threads: parseInt(this.threads),
				abs: parseInt(this.abs),
				tc: parseInt(this.tc),
				stability_aid: this.stability_aid,
				auto_clutch: this.auto_clutch,
				tyre_blankets: this.tyre_blankets,
				force_virtual_mirror: this.force_virtual_mirror,
				fuel_rate: parseInt(this.fuel_rate),
				damage_rate: parseInt(this.damage_rate),
				tires_wear_rate: parseInt(this.tires_wear_rate),
				allowed_tires_out: parseInt(this.allowed_tires_out),
				max_ballast: parseInt(this.max_ballast),
				disable_gas_cut_penality: this.disable_gas_cut_penality,
				result_screen_time: parseInt(this.result_screen_time),
				dynamic_track: this.dynamic_track,
				condition: this.condition,
				start_value: parseInt(this.start_value),
				randomness: parseInt(this.randomness),
				transferred_grip: parseInt(this.transferred_grip),
				laps_to_improve_grip: parseInt(this.laps_to_improve_grip),
				kick_vote_quorum: parseInt(this.kick_vote_quorum),
				session_vote_quorum: parseInt(this.session_vote_quorum),
				vote_duration: parseInt(this.vote_duration),
				blacklist: parseInt(this.blacklist),
				max_collisions_km: parseInt(this.max_collisions_km),
				booking: this.booking,
				booking_time: parseInt(this.booking_time),
				practice: this.practice,
				practice_time: parseInt(this.practice_time),
				can_join_practice: this.can_join_practice,
				qualify: this.qualify,
				qualify_time: parseInt(this.qualify_time),
				can_join_qualify: this.can_join_qualify,
				race: this.race,
				race_laps: parseInt(this.race_laps),
				race_time: parseInt(this.race_time),
				race_wait_time: parseInt(this.race_wait_time),
				race_extra_lap: this.race_extra_lap,
				join_type: parseInt(this.join_type),
				time: this.time,
				sun_angle: parseInt(this.sun_angle),
				weather: this.weather,
				track: this.track.name,
				track_config: this.track.config,
				legal_tyres: this.legal_tyres,
				udp_plugin_local_port: parseInt(this.udp_plugin_local_port),
				udp_plugin_address: this.udp_plugin_address,
				race_pit_window_start: parseInt(this.race_pit_window_start),
				race_pit_window_end: parseInt(this.race_pit_window_end),
				reversed_grid_race_positions: parseInt(this.reversed_grid_race_positions),
				cars: this.selectedCars
			};

			this.$http.post('/api/configuration', data)
			.then(function(resp){
				if(resp.data.code){
					console.log(resp.data.code+': '+resp.data.msg);
					this.err = resp.data.code;
					return;
				}

				this._reset();
				this._load();
				this.saved = true;
			});
		},
		performRemoveConfig: function(){
			this.$http.delete('/api/configuration', {params: {id: this._id}})
			.then(function(resp){
				if(resp.data.code){
					console.log(resp.data.code+': '+resp.data.msg);
					this.err = resp.data.code;
					return;
				}

				this._reset();
				this._load();
				this.removed = true;
			});
		},
		addWeather: function(){
			this.weather.push({
				weather: 'Clear',
				base_ambient_temp: 20,
				realistic_road_temp: 1,
				base_road_temp: 18,
				ambient_variation: 1,
				road_variation: 1,
				wind_base_speed_min: 0,
				wind_base_speed_max: 0,
				wind_base_direction: 0,
				wind_variation_direction: 0
			});
		},
		removeWeather: function(i){
			this.weather.splice(i, 1);
		},
		selectTrack: function(i){
			this.selectedTrack = i;
			this.track = this.tracks[i];
		},
		selectCar: function(i){
			this.selectedCar = i;
			this.selectedPainting = 0;
			this.activePaintings = this.cars[i].paintings;
		},
		selectPainting: function(i){
			this.selectedPainting = i;
		},
		addCar: function(){
			var car = this.cars[this.selectedCar];
			this.selectedCars.push({
				car: car.name,
				painting: car.paintings[this.selectedPainting],
				spectator: this.spectator,
				driver: this.driver,
				team: this.team,
				guid: this.guid,
				position: this.selectedCars.length,
				fixed_setup: this.fixed_setup
			});

			// only reset driver and GUID in case user wants to add multiple similar slots
			this.driver = '';
			this.guid = '';
		},
		carUp: function(i){
			if(i == 0){
				return;
			}

			var car = this.selectedCars[i-1];
			Vue.set(this.selectedCars, i-1, this.selectedCars[i]);
			Vue.set(this.selectedCars, i, car);
		},
		carDown: function(i){
			if(i == this.selectedCars.length-1){
				return;
			}

			var car = this.selectedCars[i+1];
			Vue.set(this.selectedCars, i+1, this.selectedCars[i]);
			Vue.set(this.selectedCars, i, car);
		},
		removeCar: function(i){
			this.selectedCars.splice(i, 1);
		},
		populateDynamicTrackWithPreset: function(preset) {
			switch(preset) {
				case 'DUSTY':
					this.start_value = 86;
					this.randomness = 1;
					this.transferred_grip = 50;
					this.laps_to_improve_grip = 30;
					break;
				case 'OLD':
					this.start_value = 89;
					this.randomness = 3;
					this.transferred_grip = 80;
					this.laps_to_improve_grip = 50;
					break;
				case 'SLOW':
					this.start_value = 96;
					this.randomness = 1;
					this.transferred_grip = 80;
					this.laps_to_improve_grip = 300;
					break;
				case 'GREEN':
					this.start_value = 95;
					this.randomness = 2;
					this.transferred_grip = 90;
					this.laps_to_improve_grip = 132;
					break;
				case 'FAST':
					this.start_value = 98;
					this.randomness = 2;
					this.transferred_grip = 80;
					this.laps_to_improve_grip = 700;
					break;
				case 'OPTIMUM':
					this.start_value = 100;
					this.randomness = 0;
					this.transferred_grip = 100;
					this.laps_to_improve_grip = 1;
					break;
			}
		},
		generateCfgDownloadUrl: function (id) {
			return '/api/configuration?id=' + id + '&dl=1';
		}
	}
});

File: model/configuration.go
package model

import (
	"errors"

	"github.com/jmoiron/sqlx"
)

type Configuration struct {
	Id                    int64  `json:"id"`
	Name                  string `json:"name"`
	Pwd                   string `json:"pwd"`
	AdminPwd              string `db:"admin_pwd" json:"admin_pwd"`
	PickupMode            bool   `db:"pickup_mode" json:"pickup_mode"`
	LockEntryList         bool   `db:"lock_entry_list" json:"lock_entry_list"`
	RaceOvertime          int    `db:"race_overtime" json:"race_overtime"`
	MaxSlots              int    `db:"max_slots" json:"max_slots"`
	Welcome               string `json:"welcome"`
	Description           string `json:"description"`
	UDP                   int    `json:"udp"`
	TCP                   int    `json:"tcp"`
	HTTP                  int    `json:"http"`
	PacketsHz             int    `db:"packets_hz" json:"packets_hz"`
	LoopMode              bool   `db:"loop_mode" json:"loop_mode"`
	ShowInLobby           bool   `db:"show_in_lobby" json:"show_in_lobby"`
	Threads               int    `json:"threads"`
	AuthPluginAddress     int    `db:"auth_plugin_address" json:"auth_plugin_address"`
	ABS                   int    `json:"abs"`
	TC                    int    `json:"tc"`
	StabilityAid          bool   `db:"stability_aid" json:"stability_aid"`
	AutoClutch            bool   `db:"auto_clutch" json:"auto_clutch"`
	TyreBlankets          bool   `db:"tyre_blankets" json:"tyre_blankets"`
	ForceVirtualMirror    bool   `db:"force_virtual_mirror" json:"force_virtual_mirror"`
	FuelRate              int    `db:"fuel_rate" json:"fuel_rate"`
	DamageRate            int    `db:"damage_rate" json:"damage_rate"`
	TiresWearRate         int    `db:"tires_wear_rate" json:"tires_wear_rate"`
	AllowedTiresOut       int    `db:"allowed_tires_out" json:"allowed_tires_out"`
	MaxBallast            int    `db:"max_ballast" json:"max_ballast"`
	StartRule             int    `db:"start_rule" json:"start_rule"`
	DisableGasCutPenality bool   `db:"disable_gas_cut_penality" json:"disable_gas_cut_penality"`
	TimeOfDayMult         int    `db:"time_of_day_mult" json:"time_of_day_mult"`
	ResultScreenTime      int    `db:"result_screen_time" json:"result_screen_time"`
	DynamicTrack          bool   `db:"dynamic_track" json:"dynamic_track"`
	Condition             string `db:"track_condition" json:"condition"`
	StartValue            int    `db:"start_value" json:"start_value"`
	Randomness            int    `json:"randomness"`
	TransferredGrip       int    `db:"transferred_grip" json:"transferred_grip"`
	LapsToImproveGrip     int    `db:"laps_to_improve_grip" json:"laps_to_improve_grip"`
	KickVoteQuorum        int    `db:"kick_vote_quorum" json:"kick_vote_quorum"`
	SessionVoteQuorum     int    `db:"session_vote_quorum" json:"session_vote_quorum"`
	VoteDuration          int    `db:"vote_duration" json:"vote_duration"`
	Blacklist             int    `json:"blacklist"`
	MaxCollisionsKm       int    `db:"max_collisions_km" json:"max_collisions_km"`
	Booking               bool   `json:"booking"`
	BookingTime           int    `db:"booking_time" json:"booking_time"`
	Practice              bool   `json:"practice"`
	PracticeTime          int    `db:"practice_time" json:"practice_time"`
	CanJoinPractice       bool   `db:"can_join_practice" json:"can_join_practice"`
	Qualify               bool   `json:"qualify"`
	QualifyTime           int    `db:"qualify_time" json:"qualify_time"`
	CanJoinQualify        bool   `db:"can_join_qualify" json:"can_join_qualify"`
	Race                  bool   `json:"race"`
	RaceLaps              int    `db:"race_laps" json:"race_laps"`
	RaceTime              int    `db:"race_time" json:"race_time"`
	RaceWaitTime          int    `db:"race_wait_time" json:"race_wait_time"`
	RaceExtraLap          bool   `db:"race_extra_lap" json:"race_extra_lap"`
	JoinType              int    `db:"join_type" json:"join_type"`
	SunAngle              int    `db:"sun_angle" json:"sun_angle"`
	Track                 string `json:"track"`
	TrackConfig           string `db:"track_config" json:"track_config"`
	LegalTyres            string `db:"legal_tyres" json:"legal_tyres"`
	UdpPluginPort         int    `db:"udp_plugin_local_port" json:"udp_plugin_local_port"`
	UdpPluginAddr         string `db:"udp_plugin_address" json:"udp_plugin_address"`
	RacePitWindowStart    int    `db:"race_pit_window_start" json:"race_pit_window_start"`
	RacePitWindowEnd      int    `db:"race_pit_window_end" json:"race_pit_window_end"`
	ReversedGridRacePos   int    `db:"reversed_grid_race_positions" json:"reversed_grid_race_positions"`
	ServerCfgIni          string `db:"server_cfg_ini" json:"server_cfg_ini"`
	EntryListIni          string `db:"entry_list_ini" json:"entry_list_ini"`

	Weather []Weather `db:"-" json:"weather"`
	Cars    []Car     `db:"-" json:"cars"`
}

type Weather struct {
	Id                     int64  `json:"id"`
	Configuration          int64  `json:"configuration"`
	Weather                string `json:"weather"`
	BaseAmbientTemp        int    `db:"base_ambient_temp" json:"base_ambient_temp"`
	BaseRoadTemp           int    `db:"base_road_temp" json:"base_road_temp"`
	AmbientVariation       int    `db:"ambient_variation" json:"ambient_variation"`
	RoadVariation          int    `db:"road_variation" json:"road_variation"`
	WindBaseSpeedMin       int    `db:"wind_base_speed_min" json:"wind_base_speed_min"`
	WindBaseSpeedMax       int    `db:"wind_base_speed_max" json:"wind_base_speed_max"`
	WindBaseDirection      int    `db:"wind_base_direction" json:"wind_base_direction"`
	WindVariationDirection int    `db:"wind_variation_direction" json:"wind_variation_direction"`
}

type Car struct {
	Id            int64  `json:"id"`
	Configuration int64  `json:"configuration"`
	Car           string `json:"car"`
	Painting      string `json:"painting"`
	Spectator     bool   `json:"spectator"`
	Driver        string `json:"driver"`
	Team          string `json:"team"`
	GUID          string `json:"guid"`
	Position      int    `json:"position"`
	FixedSetup    string `db:"fixed_setup" json:"fixed_setup"`
}

// Joins weather and cars.
func (m *Configuration) Join() error {
	weather, err := GetWeatherByConfiguration(m.Id)

	if err != nil {
		return err
	}

	cars, err := GetCarsByConfiguration(m.Id)

	if err != nil {
		return err
	}

	m.Weather = weather
	m.Cars = cars

	return nil
}

func (m *Configuration) Save() error {
	tx, err := session.Beginx()

	if err != nil {
		return err
	}

	// get an ID first
	if err := m.saveConfiguration(tx); err != nil {
		return err
	}

	// save weather and cars
	if err := m.saveWeather(tx); err != nil {
		return err
	}

	if err := m.saveCars(tx); err != nil {
		return err
	}

	return tx.Commit()
}

func (m *Configuration) saveConfiguration(tx *sqlx.Tx) error {
	if m.Id == 0 {
		if GetDBType() == "mysql" {
			res, err := tx.NamedExec(mysql_configuration_save, m)

			if err != nil {
				tx.Rollback()
				return err
			}

			id, err := res.LastInsertId()

			if err != nil {
				tx.Rollback()
				return err
			}

			m.Id = id
		} else {
			rows, err := tx.NamedQuery(postgres_configuration_save, m)

			if err != nil {
				tx.Rollback()
				return err
			}

			if rows.Next() {
				rows.Scan(&m.Id)
			}

			rows.Close()
		}

		return nil
	}

	var err error

	if GetDBType() == "mysql" {
		_, err = tx.NamedExec(mysql_configuration_update, m)
	} else {
		_, err = tx.NamedExec(postgres_configuration_update, m)
	}

	if err != nil {
		tx.Rollback()
		return err
	}

	return nil
}

func (m *Configuration) saveWeather(tx *sqlx.Tx) error {
	for _, weather := range m.Weather {
		if weather.Id == 0 {
			var err error

			if GetDBType() == "mysql" {
				_, err = tx.Exec(mysql_weather_save,
					m.Id,
					weather.Weather,
					weather.BaseAmbientTemp,
					weather.BaseRoadTemp,
					weather.AmbientVariation,
					weather.RoadVariation,
					weather.WindBaseSpeedMin,
					weather.WindBaseSpeedMax,
					weather.WindBaseDirection,
					weather.WindVariationDirection)
			} else {
				_, err = tx.Exec(postgres_weather_save,
					m.Id,
					weather.Weather,
					weather.BaseAmbientTemp,
					weather.BaseRoadTemp,
					weather.AmbientVariation,
					weather.RoadVariation,
					weather.WindBaseSpeedMin,
					weather.WindBaseSpeedMax,
					weather.WindBaseDirection,
					weather.WindVariationDirection)
			}

			if err != nil {
				tx.Rollback()
				return err
			}
		} else {
			var err error

			if GetDBType() == "mysql" {
				_, err = tx.Exec(mysql_weather_update,
					weather.Weather,
					weather.BaseAmbientTemp,
					weather.BaseRoadTemp,
					weather.AmbientVariation,
					weather.RoadVariation,
					weather.WindBaseSpeedMin,
					weather.WindBaseSpeedMax,
					weather.WindBaseDirection,
					weather.WindVariationDirection,
					weather.Id)
			} else {
				_, err = tx.Exec(postgres_weather_update,
					weather.Weather,
					weather.BaseAmbientTemp,
					weather.BaseRoadTemp,
					weather.AmbientVariation,
					weather.RoadVariation,
					weather.WindBaseSpeedMin,
					weather.WindBaseSpeedMax,
					weather.WindBaseDirection,
					weather.WindVariationDirection,
					weather.Id)
			}

			if err != nil {
				tx.Rollback()
				return err
			}
		}
	}

	return nil
}

func (m *Configuration) saveCars(tx *sqlx.Tx) error {
	for _, car := range m.Cars {
		if car.Id == 0 {
			var err error

			if GetDBType() == "mysql" {
				_, err = tx.Exec(mysql_cars_save,
					m.Id,
					car.Car,
					car.Painting,
					car.Spectator,
					car.Driver,
					car.Team,
					car.GUID,
					car.Position,
					car.FixedSetup)
			} else {
				_, err = tx.Exec(postgres_cars_save,
					m.Id,
					car.Car,
					car.Painting,
					car.Spectator,
					car.Driver,
					car.Team,
					car.GUID,
					car.Position,
					car.FixedSetup)
			}

			if err != nil {
				tx.Rollback()
				return err
			}
		} else {
			var err error

			if GetDBType() == "mysql" {
				_, err = tx.Exec(mysql_cars_update,
					car.Car,
					car.Painting,
					car.Spectator,
					car.Driver,
					car.Team,
					car.GUID,
					car.Position,
					car.FixedSetup,
					car.Id)
			} else {
				_, err = tx.Exec(postgres_cars_update,
					car.Car,
					car.Painting,
					car.Spectator,
					car.Driver,
					car.Team,
					car.GUID,
					car.Position,
					car.FixedSetup,
					car.Id)
			}

			if err != nil {
				tx.Rollback()
				return err
			}
		}
	}

	return nil
}

func (m *Configuration) Remove() error {
	if m.Id == 0 {
		return errors.New("ID must be set")
	}

	tx, err := session.Beginx()

	if err != nil {
		return err
	}

	if GetDBType() == "mysql" {
		_, err = tx.NamedExec(mysql_weather_delete_configuration, m)

		if err != nil {
			tx.Rollback()
			return err
		}

		_, err = tx.NamedExec(mysql_cars_delete_configuration, m)

		if err != nil {
			tx.Rollback()
			return err
		}

		_, err = tx.NamedExec(mysql_configuration_delete, m)

		if err != nil {
			tx.Rollback()
			return err
		}
	} else {
		_, err = tx.NamedExec(postgres_weather_delete_configuration, m)

		if err != nil {
			tx.Rollback()
			return err
		}

		_, err = tx.NamedExec(postgres_cars_delete_configuration, m)

		if err != nil {
			tx.Rollback()
			return err
		}

		_, err = tx.NamedExec(postgres_configuration_delete, m)

		if err != nil {
			tx.Rollback()
			return err
		}
	}

	return tx.Commit()
}

func (m *Weather) Remove() error {
	var err error

	if GetDBType() == "mysql" {
		_, err = session.NamedExec(mysql_weather_delete, m)
	} else {
		_, err = session.NamedExec(postgres_weather_delete, m)
	}

	return err
}

func (m *Car) Remove() error {
	var err error

	if GetDBType() == "mysql" {
		_, err = session.NamedExec(mysql_cars_delete, m)
	} else {
		_, err = session.NamedExec(postgres_cars_delete, m)
	}

	return err
}

func GetWeatherByConfiguration(id int64) ([]Weather, error) {
	weather := make([]Weather, 0)

	if GetDBType() == "mysql" {
		if err := session.Select(&weather, mysql_weather_get_configuration, id); err != nil {
			return nil, err
		}
	} else {
		if err := session.Select(&weather, postgres_weather_get_configuration, id); err != nil {
			return nil, err
		}
	}

	return weather, nil
}

func GetCarsByConfiguration(id int64) ([]Car, error) {
	cars := make([]Car, 0)

	if GetDBType() == "mysql" {
		if err := session.Select(&cars, mysql_cars_get_configuration, id); err != nil {
			return nil, err
		}
	} else {
		if err := session.Select(&cars, postgres_cars_get_configuration, id); err != nil {
			return nil, err
		}
	}

	return cars, nil
}

func GetAllConfigurations() ([]Configuration, error) {
	configs := make([]Configuration, 0)

	if GetDBType() == "mysql" {
		if err := session.Select(&configs, mysql_configuration_get_all); err != nil {
			return nil, err
		}
	} else {
		if err := session.Select(&configs, postgres_configuration_get_all); err != nil {
			return nil, err
		}
	}

	return configs, nil
}

func GetConfigurationById(id int64) (*Configuration, error) {
	config := new(Configuration)

	if GetDBType() == "mysql" {
		if err := session.Get(config, mysql_configuration_get_id, id); err != nil {
			return nil, err
		}
	} else {
		if err := session.Get(config, postgres_configuration_get_id, id); err != nil {
			return nil, err
		}
	}

	return config, nil
}
